From 4f2db567102772472686a28f44cef41e051b15b7 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Tarmo=20T=C3=A4nav?= <tarmo@itech.ee>
Date: Thu, 3 Jul 2008 17:49:12 +0300
Subject: [PATCH] Use :namespace instead of :path_prefix for finding controller

:namespace is supposed to be the module where controller exists.
:path_prefix can contain anything, including variables, which
makes it unsuitable for determining the module for a controller.
---
 .../lib/action_controller/routing/builder.rb       |    3 +--
 actionpack/test/controller/routing_test.rb         |   20 ++++++++++++++++++++
 2 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/actionpack/lib/action_controller/routing/builder.rb b/actionpack/lib/action_controller/routing/builder.rb
index 4740113..b832384 100644
--- a/actionpack/lib/action_controller/routing/builder.rb
+++ b/actionpack/lib/action_controller/routing/builder.rb
@@ -67,10 +67,9 @@ module ActionController
         options = options.dup
 
         if options[:namespace]
-          options[:controller] = "#{options[:path_prefix]}/#{options[:controller]}"
+          options[:controller] = "#{options.delete(:namespace).sub(/\/$/, '')}/#{options[:controller]}"
           options.delete(:path_prefix)
           options.delete(:name_prefix)
-          options.delete(:namespace)
         end
 
         requirements = (options.delete(:requirements) || {}).dup
diff --git a/actionpack/test/controller/routing_test.rb b/actionpack/test/controller/routing_test.rb
index 07c13eb..c5ccb71 100644
--- a/actionpack/test/controller/routing_test.rb
+++ b/actionpack/test/controller/routing_test.rb
@@ -2039,6 +2039,26 @@ uses_mocha 'LegacyRouteSet, Route, RouteSet and RouteLoading' do
       Object.send(:remove_const, :Api)
     end
 
+    def test_namespace_with_path_prefix
+      Object.const_set(:Api, Module.new { |m| m.const_set(:ProductsController, Class.new) })
+
+      set.draw do |map|
+
+        map.namespace 'api', :path_prefix => 'prefix' do |api|
+          api.route 'inventory', :controller => "products", :action => 'inventory'
+        end
+
+      end
+
+      request.path = "/prefix/inventory"
+      request.method = :get
+      assert_nothing_raised { set.recognize(request) }
+      assert_equal("api/products", request.path_parameters[:controller])
+      assert_equal("inventory", request.path_parameters[:action])
+    ensure
+      Object.send(:remove_const, :Api)
+    end
+
     def test_generate_finds_best_fit
       set.draw do |map|
         map.connect "/people", :controller => "people", :action => "index"
-- 
1.5.5.4

